CRM/Core/Form.php

```diff
@@ -1109,6 +1109,142 @@ public function addSelect($name, $props = array(), $required = FALSE) {
+    foreach ($info['values'] as $uniqueName => $fieldSpec) {
+      if (
+          $uniqueName === $props['field'] ||
+          CRM_Utils_Array::value('name', $fieldSpec) === $props['field'] ||
+          in_array($props['field'], CRM_Utils_Array::value('api.aliases', $fieldSpec, array()))
+      ) {
+        break;
+      }
+    }
+
+    $label = isset($props['label']) ? $props['label'] : $fieldSpec['title'];
+
+    if (!array_key_exists('placeholder', $props)) {
+      $props['placeholder'] = $required ? ts('- select -') : CRM_Utils_Array::value('context', $props) == 'search' ? ts('- any -') : ts('- none -');
+    }
+
+    $widget = isset($props['type']) ? $props['type'] : $fieldSpec['html']['type'];
+
+    if ($widget == 'TextArea' && CRM_Utils_Array::value('context', $props) == 'search') {
+      $widget = 'Text';
+    }
+
+    $isSelect = (in_array($widget, array(
+          'Select',
+          'Multi-Select',
+          'Select State/Province',
+          'Multi-Select State/Province',
+          'Select Country',
+          'Multi-Select Country',
+          'AdvMulti-Select',
+          'CheckBox',
+          'Radio',
+    )));
+
+    if ($isSelect) {
+      // Fetch options from the api unless passed explicitly
+      if (isset($props['options'])) {
+        $options = $props['options'];
+      }
+      else {
+        $info = civicrm_api3($props['entity'], 'getoptions', $props);
+        $options = $info['values'];
+      }
+      if (CRM_Utils_Array::value('context', $props) == 'search' || ($widget !== 'AdvMulti-Select' && strpos($widget, 'Select') !== FALSE)) {
+        $widget = 'Select';
+      }
+      $props['class'] = (isset($props['class']) ? $props['class'] . ' ' : '') . "crm-select2";
+      if (CRM_Utils_Array::value('context', $props) == 'search' || strpos($widget, 'Multi') !== FALSE) {
+        $props['class'] .= ' huge';
+        $props['multiple'] = 'multiple';
+      }
+
+      // Add data for popup link.
+      if (CRM_Utils_Array::value('context', $props) != 'search' && $widget == 'Select' && CRM_Core_Permission::check('administer CiviCRM')) {
+        $props['data-option-edit-path'] = array_key_exists('option_url', $props) ? $props['option_url'] : $props['data-option-edit-path'] = CRM_Core_PseudoConstant::getOptionEditUrl($fieldSpec);
+        $props['data-api-entity'] = $props['entity'];
+        $props['data-api-field'] = $props['field'];
+      }
+    }
+
+    switch ($widget) {
+      case 'Text':
+      case 'Link':
+        //TODO: Autodetect ranges
+        $this->add('text', $name, $label, NULL, $required, $props);
+        break;
+
+      case 'TextArea':
+        //Set default rows and cols
+        $props['rows'] = isset($props['rows']) ? $props['rows'] : 4;
+        $props['cols'] = isset($props['cols']) ? $props['cols'] : 60;
+        $this->add('textarea', $name, $label, NULL, $required, $props);
+        break;
+
+      case 'Select Date':
+        //TODO: Add date formats
```

Comment:
Yea another todo is to remove the dependency of manually including the datepicker js in the template.
I've got a separate issue open for that somewhere...